### PR TITLE
Offline Compilation Support for Animations

### PR DESCRIPTION
**Pull request by matsko**

### CODE REVIEW

File: modules/@angular/compiler-cli/integrationtest/src/animate.ts [hunk 1]
```diff
@@ -0,0 +1,46 @@
+import {Component, trigger, state, animate, transition, style} from '@angular/core';
+
+@Component({
+  selector: "animate-cmp",
+  styles: [`
```
Comment (tbosch) [inline modules/@angular/compiler-cli/integrationtest/src/animate.ts]: Do you need these for the test?

File: modules/@angular/compiler-cli/integrationtest/test/animate_spec.ts [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+describe("template codegen output", () => {
+  it("should apply the animate states to the element", (done) => {
+    const appInjector = ReflectiveInjector.resolveAndCreate(BROWSER_APP_PROVIDERS,
+      browserPlatform().injector);
+    var comp = AnimateCmpNgFactory.create(appInjector);
+    var debugElement = <DebugElement>getDebugNode(comp.location.nativeElement);
+
+    // the open-close-container is a child of the main container
+    // if the template changes then please update the location below
+    var targetDebugElement = <DebugElement>debugElement.childNodes[7];
```
Comment (tbosch) [inline modules/@angular/compiler-cli/integrationtest/test/animate_spec.ts]: Use `.children` which will ignore text nodes.

File: modules/@angular/core/src/debug/debug_renderer.ts [hunk 3]
```diff
@@ -125,14 +127,29 @@ export class DebugDomRenderer implements Renderer {
   }
 
   setElementClass(renderElement: any, className: string, isAdd: boolean) {
+    var debugEl = getDebugNode(renderElement);
+    if (isPresent(debugEl) && debugEl instanceof DebugElement) {
+      debugEl.classes[className] = isAdd ? className : null;
```
Comment (tbosch) [inline modules/@angular/core/src/debug/debug_renderer.ts]: The value should always be `isAdd`, i.e. a boolean.

File: modules/@angular/core/src/debug/debug_node.ts [hunk 4]
```diff
@@ -56,13 +56,17 @@ export class DebugElement extends DebugNode {
   name: string;
   properties: {[key: string]: string};
   attributes: {[key: string]: string};
+  classes: {[key: string]: string};
```
Comment (tbosch) [inline modules/@angular/core/src/debug/debug_node.ts]: This should be a `{[key:string]:boolean}`.

File: modules/@angular/core/src/debug/debug_renderer.ts [hunk 5]
```diff
@@ -125,14 +127,29 @@ export class DebugDomRenderer implements Renderer {
...
+    }
     this._delegate.setElementClass(renderElement, className, isAdd);
   }
 
   setElementStyles(renderElement: any, styles: {[key: string]: string}) {
```
Comment (tbosch) [inline modules/@angular/core/src/debug/debug_renderer.ts]: Can you add a second commit that removes this method from the `Renderer`? 
It is just looping over the styles and then does the same as `setElementStyle`. Just add a helper function in `view_utils.ts` if you need this for codegen that that takes in the map and the renderer and calls `setElementStyle` individually.

Background: We don't have these collection calls for other things like classes either.

File: modules/@angular/core/src/debug/debug_renderer.ts [hunk 6]
```diff
@@ -146,3 +163,6 @@ export class DebugDomRenderer implements Renderer {
     return this._delegate.animate(element, startingStyles, keyframes, duration, delay, easing);
   }
 }
+
+function _updateDebugElementStyles(debugEl: DebugElement, styles: {[key: string]: string}) {
```
Comment (tbosch) [inline modules/@angular/core/src/debug/debug_renderer.ts]: Remove this

File: modules/@angular/core/src/debug/debug_renderer.ts (hunk 5 above)
Comment (tbosch) [inline modules/@angular/core/src/debug/debug_renderer.ts]: Or better in `animation_utils`.

File: modules/@angular/core/test/debug/debug_node_spec.ts [hunk 7]
```diff
@@ -285,6 +293,30 @@ export function main() {
          });
        }));
 
+    it('should list element classes',
+       inject([TestComponentBuilder, AsyncTestCompleter], (tcb: TestComponentBuilder, async) => {
+         tcb.createAsync(TestApp).then((fixture) => {
+           fixture.detectChanges();
+           var bankElem = fixture.debugElement.children[0];
+
+           expect(bankElem.classes['closed']).toBeTruthy();
```
Comment (tbosch) [inline modules/@angular/core/test/debug/debug_node_spec.ts]: `toBe(true)` and `toBe(false)`.

**Comment by matsko** [comment]
Landed as fa0718ba9ac53c5770e2d8b7809d7d9aca6ee852

**Pull request closed without merging**